array_api_tests/dtype_helpers.py

```diff
@@ -0,0 +1,289 @@
+from . import _array_module as xp
+
+
+__all__ = [
+    'dtypes_to_scalars',
+    'category_to_dtypes',
+    'promotion_table',
+    'dtype_nbits',
+    'dtype_signed',
+    'func_in_categories',
+    'func_out_categories',
+    'op_in_categories',
+    'op_out_categories',
+    'binary_op_to_symbol',
+    'unary_op_to_symbol',
+    'op_to_func',
+]
+
+
+int_dtypes = (xp.int8, xp.int16, xp.int32, xp.int64)
+uint_dtypes = (xp.uint8, xp.uint16, xp.uint32, xp.uint64)
+all_int_dtypes = int_dtypes + uint_dtypes
+float_dtypes = (xp.float32, xp.float64)
+numeric_dtypes = all_int_dtypes + float_dtypes
+all_dtypes = (xp.bool,) + numeric_dtypes
+
+
+dtypes_to_scalars = {
+    xp.bool: [bool],
+    **{d: [int] for d in all_int_dtypes},
+    **{d: [int, float] for d in float_dtypes},
+}
+
+
+category_to_dtypes = {
+    'any': all_dtypes,
+    'boolean': (xp.bool,),
+    'floating': float_dtypes,
+    'integer': all_int_dtypes,
+    'integer_or_boolean': (xp.bool,) + uint_dtypes + int_dtypes,
+    'numeric': numeric_dtypes,
+}
+
+
+_numeric_promotions = {
+    # ints
+    (xp.int8, xp.int8): xp.int8,
+    (xp.int8, xp.int16): xp.int16,
+    (xp.int8, xp.int32): xp.int32,
+    (xp.int8, xp.int64): xp.int64,
+    (xp.int16, xp.int16): xp.int16,
+    (xp.int16, xp.int32): xp.int32,
+    (xp.int16, xp.int64): xp.int64,
+    (xp.int32, xp.int32): xp.int32,
+    (xp.int32, xp.int64): xp.int64,
+    (xp.int64, xp.int64): xp.int64,
+    # uints
+    (xp.uint8, xp.uint8): xp.uint8,
+    (xp.uint8, xp.uint16): xp.uint16,
+    (xp.uint8, xp.uint32): xp.uint32,
+    (xp.uint8, xp.uint64): xp.uint64,
+    (xp.uint16, xp.uint16): xp.uint16,
+    (xp.uint16, xp.uint32): xp.uint32,
+    (xp.uint16, xp.uint64): xp.uint64,
+    (xp.uint32, xp.uint32): xp.uint32,
+    (xp.uint32, xp.uint64): xp.uint64,
+    (xp.uint64, xp.uint64): xp.uint64,
+    # ints and uints (mixed sign)
+    (xp.int8, xp.uint8): xp.int16,
+    (xp.int8, xp.uint16): xp.int32,
+    (xp.int8, xp.uint32): xp.int64,
+    (xp.int16, xp.uint8): xp.int16,
+    (xp.int16, xp.uint16): xp.int32,
+    (xp.int16, xp.uint32): xp.int64,
+    (xp.int32, xp.uint8): xp.int32,
+    (xp.int32, xp.uint16): xp.int32,
+    (xp.int32, xp.uint32): xp.int64,
+    (xp.int64, xp.uint8): xp.int64,
+    (xp.int64, xp.uint16): xp.int64,
+    (xp.int64, xp.uint32): xp.int64,
+    # floats
+    (xp.float32, xp.float32): xp.float32,
+    (xp.float32, xp.float64): xp.float64,
+    (xp.float64, xp.float64): xp.float64,
+}
+promotion_table = {
+    (xp.bool, xp.bool): xp.bool,
+    **_numeric_promotions,
+    # TODO: dont unpack pairs of the same dtype
```

Comment:
This doesn't matter. The dict will already de-duplicate. The only way this might matter is if we want to make the order reasonable, but in that case the below dict should be mixed with this one.